### PR TITLE
end2endtest: add ballot test

### PR DESCRIPTION
**Pull request by mariajdab**
- #863

### CODE REVIEW

**Comment by p4u** [comment]
please @altergui review

**Comment by p4u** [comment]
We need to find out why the CI is failing @altergui @mariajdab

**Comment by altergui** [comment]
> We need to find out why the CI is failing @altergui @mariajdab

it's timing out in the middle of plaintext election :thinking: taking more than 3-4 minutes to generate 100 votes.
this PR doesn't touch plaintext election test at all. so my only guess is that this new ballot test is overwhelming the network, is it resource-intensive in any way? if not, (it's as light as the other tests) it might be simply that the CI testsuite can't handle so many tests concurrently.

File: cmd/end2endtest/ballot.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+func (t *E2EBallotElection) Run() error {
+	c := t.config
+	api := t.api
+
+	voteValues, expectedR := t.randomBallotVotes()
+
+	// convert expected results int slice to string slice to be used on compare the results later
+	time.Sleep(time.Second * 6)
```
Comment (altergui) [inline cmd/end2endtest/ballot.go]: why wait here?

Comment (mariajdab) [reply cmd/end2endtest/ballot.go]: To have some time to return the `expectedR` before use it in the goroutine below, but seems is not necessary

Comment (altergui) [reply cmd/end2endtest/ballot.go]: `expectedR` is returned before the sleep even starts. no need to wait

**Comment by p4u** [comment]
@altergui please review again and either accept it or request more changes.

@mariajdab you will need to merge with existing master and fix the conflicts.

**Comment by mariajdab** [comment]
@altergui this CI shows the error that I mentioned you previously. I think the results expected are Ok. The code is using `WaitUntilElectionResults`

**Comment by altergui** [comment]
i looked at the logs and i conclude there's one vote that you expect to be discarded, it's not being discarded
```
VoteMode:uniqueValues:true 
TallyMode:maxCount:2 
          maxValue:6
          maxTotalCost:10
```
```
vote values generated[[1 1] [2 6] [4 0] [1 6] [8 4] [6 5] [1 3] [0 6] [6 3] [3 3]]
results expected[[10 20 10 0 10 0 10] [10 0 0 20 0 0 30]]
```

you expect 4 votes to be discarded: 
 * `[1 1]` (fails uniqueValues)
 * `[8 4]` (8 + 4 = 12 is more than maxTotalCost=10)
 * `[6 5]` (6 + 5 = 11 is more than maxTotalCost=10)
 * `[3 3]` (fails uniqueValues)
 
 and yet,
 ```
 got Results: [[10 20 10 0 10 0 20] [10 0 0 20 0 10 30]]
 ```
 
 those results indicate that the `[6 5]` vote was not discarded, it was counted as valid.
 
 so there's "off by one" bug somewhere

**Comment by altergui** [comment]
or, maxTotalCost is not being checked at all.
and the [8 4] vote failed because of maxValue

i'm still doing tests

**Comment by mariajdab** [comment]
> i looked at the logs and i conclude there's one vote that you expect to be discarded, it's not being discarded
> 
> ```
> VoteMode:uniqueValues:true 
> TallyMode:maxCount:2 
>           maxValue:6
>           maxTotalCost:10
> ```
> 
> ```
> vote values generated[[1 1] [2 6] [4 0] [1 6] [8 4] [6 5] [1 3] [0 6] [6 3] [3 3]]
> results expected[[10 20 10 0 10 0 10] [10 0 0 20 0 0 30]]
> ```
> 
> you expect 4 votes to be discarded:
> 
>     * `[1 1]` (fails uniqueValues)
> 
>     * `[8 4]` (8 + 4 = 12 is more than maxTotalCost=10)
> 
>     * `[6 5]` (6 + 5 = 11 is more than maxTotalCost=10)
> 
>     * `[3 3]` (fails uniqueValues)
> 
> 
> and yet,
> 
> ```
> got Results: [[10 20 10 0 10 0 20] [10 0 0 20 0 10 30]]
> ```
> 
> those results indicate that the `[6 5]` vote was not discarded, it was counted as valid.
> 
> so there's "off by one" bug somewhere

Yeah, agree that's actually I realized trying to test repeatedly to make sure that it truly pass and there is not 'possible error'. And this is the error that some times appears that count a vote that should not be counted in the final results.

**Comment by altergui** [comment]
indeed, the issue is that maxTotalCost is not being checked. but i believe this is a bug on the API, that is not respecting the parameter you're passing (MaxTotalCost=10) and instead defaulting to a hardcoded math that never fails. i'll post details in a bit

**Comment by altergui** [comment]
indeed. the value you're passing is not respected at all, TWICE :rofl: 
first apiclient is ignoring what you passed, and pushing a hardcoded math to the api
https://github.com/vocdoni/vocdoni-node/blob/af85c0571e2834cf85a1906e1ece99c12f8939a2/apiclient/election.go#L174
and then the api ALSO ignores what apiclient pushed, and also recalculates the same math from scratch
https://github.com/vocdoni/vocdoni-node/blob/e5f82f9c8ad680ed952e4b17d19a9b14f01fe61b/api/wallet.go#L405

**Comment by altergui** [comment]
mmh nevermind, now i see the NewElectionRaw uses a different codepath.

**Comment by altergui** [comment]
ah, found the issue. 
you are missing `CostExponent: 1,` in your VoteOptions

the default CostExponent is 0 , which means the vote `[6 5]` is computed as `6**0 + 5**0 = 2`

**Comment by mariajdab** [comment]
> ah, found the issue. you are missing `CostExponent: 1,` in your VoteOptions
> 
> the default CostExponent is 0 , which means the vote `[6 5]` is computed as `6**0 + 5**0 = 2`

Got it!

**Comment by altergui** [comment]
i'm preparing a commit, i'll push it here and if you like it just squash it with yours

File: cmd/end2endtest/helpers.go [hunk 2]
```diff
@@ -137,10 +138,21 @@ func (t *e2eElection) isCensusSizeValid(censusID types.HexBytes) bool {
...
 
-func (t *e2eElection) createElection(electionDescrip *vapi.ElectionDescription) (*vapi.Election, error) {
-	electionID, err := t.api.NewElection(electionDescrip)
-	if err != nil {
-		return nil, err
+func (t *e2eElection) createElection(ed *vapi.ElectionDescription, prc *models.Process) (*vapi.Election, error) {
+	var electionID types.HexBytes
+	var err error
+
+	// TODO: improve this
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: this really complicated debugging the MaxTotalCost issue, i didn't expect createElection take two different codepaths.

a more idiomatic alternative i can suggest is making a `func (t *e2eElection) createElectionRaw`

**Review by altergui — CHANGES_REQUESTED** [review]

File: cmd/end2endtest/helpers.go [hunk 3]
```diff
@@ -322,17 +356,17 @@ func (t *e2eElection) overwriteVote(choices []int, indexAcct int, waitType strin
...
 			if strings.Contains(err.Error(), "overwrite count reached") {
-				log.Infof("error expected: %s", err.Error())
+				log.Debugw("error expected: %s", err.Error())
 			} else {
 				return 0, errors.New("expected overwrite error")
 			}
 		}
 		contextDeadlines += ctxDeadLine
 		switch waitType {
-		case nextBlock:
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: this looks totally unrelated to "add ballot test", is this a fix for vote overwrite? in that case please make a separate PR

File: cmd/end2endtest/helpers.go [hunk 4]
```diff
@@ -306,11 +339,12 @@ func (t *e2eElection) setupElection(ed *vapi.ElectionDescription) error {
 		return err
 	}
 
-	t.proofs = t.generateProofs(root, ed.ElectionType.Anonymous)
+	t.proofs = t.generateProofs(root, anonymousVoting)
 
 	return nil
 }
 
+// overwriteVote allow to try to overwrite a previous vote given the index of the account, it can use the sameBlock or the nextBlock
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: this func comment also looks unrelated. being just a comment, it would not merit a PR, but if you're doing a PR for the overwrite fixes (sameBlock vs nextBlock) please move these changes to that PR as well.

File: cmd/end2endtest/helpers.go [hunk 5]
```diff
@@ -379,6 +414,59 @@ func (t *e2eElection) sendVote(v voteInfo, apiClientMtx *sync.Mutex) (int, error
...
 
+// randomBallotVotes allow to create random vote values that possibly exceed the max value, max total cost and is not unique
+func (t *e2eElection) randomBallotVotes() ([][]int, [][]int) {
+	maxCount := int(t.election.TallyMode.MaxCount)
+	maxValue := int(t.election.TallyMode.MaxValue)
+	maxTotalCost := int(t.election.TallyMode.MaxTotalCost)
+	costExponent := int(t.election.TallyMode.CostExponent)
+
+	expecResults := make([][]int, maxCount)
+	voteValues := make([][]int, 0)
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: for clarity, (and consistency with other parts of the code)
i'd suggest `expectedResults`

and `voteValues` could be simply `votes`

File: cmd/end2endtest/helpers.go [hunk 6]
```diff
@@ -406,13 +494,25 @@ func getCensusParticipantKey(voterAccount *ethereum.SignKeys, censusType string)
 	return key, nil
 }
 
-// matchResult compare the expected vote results base in the overwrite applied, with the actual result returned by the API
-func matchResult(results [][]*types.BigInt, expectedResult [][]string) bool {
-	// only has 1 question
-	for q := range results[0] {
-		if !(expectedResult[0][q] == results[0][q].String()) {
-			return false
+func matchResults(results [][]*types.BigInt, expectedResult [][]string) bool {
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: missing s in `expectedResults`

File: cmd/end2endtest/helpers.go [hunk 7]
```diff
@@ -406,13 +494,25 @@ func getCensusParticipantKey(voterAccount *ethereum.SignKeys, censusType string)
...
+	// iterate over each question to check if the results match with the expected results
+	for i := 0; i < len(results); i++ {
+		for q := range results[i] {
+			if !(expectedResult[i][q] == results[i][q].String()) {
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: i'd rather make expectedResults a `[][]*types.BigInt`, you're testing small ints so you don't mind but if results are BigInt let's respect that

File: cmd/end2endtest/helpers.go [hunk 8]
```diff
@@ -406,13 +494,25 @@ func getCensusParticipantKey(voterAccount *ethereum.SignKeys, censusType string)
...
+				return false
+			}
 		}
 	}
 	return true
 }
+
+func sliceIntToStr(intSlice [][]int) [][]string {
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: you also can drop this whole func if you make expectedResults a `[][]*types.BigInt`

File: cmd/end2endtest/main.go [hunk 9]
```diff
@@ -110,6 +110,10 @@ func main() {
 	flag.CommandLine.SortFlags = false
 	flag.Parse()
 
+	if c.parallelCount > c.nvotes {
+		c.parallelCount = c.nvotes
+	}
+
```
Comment (altergui) [inline cmd/end2endtest/main.go]: this is actually an unrelated fix but in this case it was me :sweat_smile: if you make a separate PR ("end2endtest small fixes") i'll move it there

Comment (altergui) [reply cmd/end2endtest/main.go]: along with voteOverwrite fixes and so on

File: dockerfiles/testsuite/start_test.sh [hunk 10]
```diff
@@ -116,14 +117,18 @@ e2etest_tokentxs() {
 	e2etest tokentxs
 }
 
-e2etest_overwritelection () {
+e2etest_overwritelection() {
   e2etest overwritelection
 }
 
-e2etest_cenususizelection () {
+e2etest_cenususizelection() {
```
Comment (altergui) [inline dockerfiles/testsuite/start_test.sh]: another candidate to be moved into "e2etest small fixes"
cenususizelection -> censusizelection

Comment (mariajdab) [reply dockerfiles/testsuite/start_test.sh]: Yeah, sure I will create a small PR for those

File: cmd/end2endtest/helpers.go (hunk 3 above)
Comment (mariajdab) [reply cmd/end2endtest/helpers.go]: yeah no problem

File: cmd/end2endtest/ballot.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		resultsChan <- sliceIntToStr(expectedR)
+	}()
+
+	results := <-resultsChan
+	wg.Wait()
```
Comment (altergui) [inline cmd/end2endtest/ballot.go]: also, this whole block looks like something you did while debugging the expectedResults bug. (i remember there was a time.Sleep here as well)
i must say it's a very peculiar way of doing `results := sliceIntToStr(expectedR)`
and as i said, you really should make `expectedResults` a slice of BigInts, so you don't even need that sliceIntToStr conversion,
and you can delete this whole thing entirely and just use `expectedR` (renamed to `expectedResults` for consistency)

**Comment by mariajdab** [comment]
@altergui Seems that there is something wrong, is not a frequent error but if you try to run around 8 times it will appears, but now the error is in `expectedResults`. I'm checking the last changes added to `randomBallotVotes` [job-link](https://github.com/vocdoni/vocdoni-node/actions/runs/5050540323/jobs/9108586183)

File: cmd/end2endtest/helpers.go (hunk 3 above)
Comment (mariajdab) [reply cmd/end2endtest/helpers.go]: PR created https://github.com/vocdoni/vocdoni-node/pull/943

**Comment by altergui** [comment]
i see you're still working on this so i'm not reviewing it, when you finish just re-request review

a quick question, why `expectedResults` vs `resultExpected` vs `expecResults`? if they are all the same concept, please keep the var name consistent.
also, small abbreviations like `expecResults` are most likely not worth the loss of readability vs 3 letters less, better keep the words complete.

**Comment by mariajdav** [comment]
> i see you're still working on this so i'm not reviewing it, when you finish just re-request review
> 
> a quick question, why `expectedResults` vs `resultExpected` vs `expecResults`? if they are all the same concept, please keep the var name consistent.
> also, small abbreviations like `expecResults` are most likely not worth the loss of readability vs 3 letters less, better keep the words complete.

Yes, I am still working on this.

**Comment by mariajdab** [comment]
> i see you're still working on this so i'm not reviewing it, when you finish just re-request review
> 
> a quick question, why `expectedResults` vs `resultExpected` vs `expecResults`? if they are all the same concept, please keep the var name consistent. also, small abbreviations like `expecResults` are most likely not worth the loss of readability vs 3 letters less, better keep the words complete.

Yes, there are some small fixes that I need to push

**Comment by mariajdab** [comment]
> > i see you're still working on this so i'm not reviewing it, when you finish just re-request review
> > a quick question, why `expectedResults` vs `resultExpected` vs `expecResults`? if they are all the same concept, please keep the var name consistent. also, small abbreviations like `expecResults` are most likely not worth the loss of readability vs 3 letters less, better keep the words complete.
> 
> Yes, there are some small fixes that I need to push

I just pushed

File: cmd/end2endtest/censusize.go [hunk 12]
```diff
@@ -154,13 +155,19 @@ func (t *E2EMaxCensusSizeElection) Run() error {
...
+	choices := []*types.BigInt{
+		// should not count the last vote
+		new(types.BigInt).SetUint64(uint64(c.nvotes-1) * 10),
+		// should count the firsts overwrite
+		new(types.BigInt).SetUint64(0),
+		new(types.BigInt).SetUint64(0),
+	}
+
+	// according to the first overwrite
+	expectedResults := [][]*types.BigInt{choices}
```
Comment (altergui) [inline cmd/end2endtest/censusize.go]: you can merge both assignments 
```go
expectedResults := [][]*types.BigInt{
  new(types.BigInt)....
 }
```

since `choices` is not used anywhere else

File: cmd/end2endtest/helpers.go [hunk 13]
```diff
@@ -393,6 +441,76 @@ func (t *e2eElection) sendVote(v voteInfo, apiClientMtx *sync.Mutex) (int, error
...
+		new(types.BigInt).SetUint64(10),
+	}
+	// default results on choice 2 for 10 votes
+	resultsCh2 := []*types.BigInt{
+		new(types.BigInt).SetUint64(10), new(types.BigInt).SetUint64(0),
+		new(types.BigInt).SetUint64(0), new(types.BigInt).SetUint64(10),
+		new(types.BigInt).SetUint64(10), new(types.BigInt).SetUint64(0),
+		new(types.BigInt).SetUint64(10),
+	}
+
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: there must be a way to make this less repetitive

File: cmd/end2endtest/helpers.go [hunk 14]
```diff
@@ -393,6 +441,76 @@ func (t *e2eElection) sendVote(v voteInfo, apiClientMtx *sync.Mutex) (int, error
 	return contextDeadline, nil
 }
 
+// ballotVotes allow to create random vote values that possibly exceed the max value, max total cost and is not unique
+func (t *e2eElection) ballotVotes() ([][]int, [][]*types.BigInt) {
```
Comment (altergui) [inline cmd/end2endtest/helpers.go]: i have the feeling this is not generic enough to belong to e2eElection, isn't it something specific of E2EBallotElection?

Comment (mariajdab) [reply cmd/end2endtest/helpers.go]: Yeah, agree. It could be just a function

File: cmd/end2endtest/helpers.go [hunk 15]
```diff
@@ -393,6 +448,76 @@ func (t *e2eElection) sendVote(v voteInfo, apiClientMtx *sync.Mutex) (int, error
...
+
+	switch {
+	// less than 10 votes
+	case nvotes < 10:
+		// default results on choice 1 for less than 10 votes
+		resultsCh1 = votesToBigInt([]uint64{0, 0, 0, 0, 0, 0, 0})
+		// default results on choice 2 for less than 10 votes
+		resultsCh2 = votesToBigInt([]uint64{0, 0, 0, 0, 0, 0, 0})
+
+	// greater o equal than 10 votes
```
Comment (altergui) [inline cmd/end2endtest/helpers.go:470]: i believe this two assignments could be de-indented (and instead of a `switch` a simple `if`)
or, even, that this whole block is redundant (since it assigns the zero value?), but i'm not sure.

File: cmd/end2endtest/helpers.go [hunk 16]
```diff
@@ -420,13 +545,22 @@ func censusParticipantKey(voterAccount *ethereum.SignKeys, censusType string) ([
...
+		for q := range results[i] {
+			if !(expectedResults[i][q].String() == results[i][q].String()) {
+				return false
+			}
 		}
 	}
 	return true
 }
+
+func votesToBigInt(votes []uint64) []*types.BigInt {
```
Comment (altergui) [inline cmd/end2endtest/helpers.go:560]: it would be much prettier for the caller if this would be 
`votesToBigInt(votes ...int)`

so the calls would be

`resultsCh1 = votesToBigInt(0, 10, 20, 0, 0, 0, 10)`

File: cmd/end2endtest/helpers.go [hunk 17]
```diff
@@ -393,6 +448,76 @@ func (t *e2eElection) sendVote(v voteInfo, apiClientMtx *sync.Mutex) (int, error
 	return contextDeadline, nil
 }
 
+// ballotVotes from a default list of 10 vote values that exceed the max value, max total cost and is not unique will
+func ballotVotes(b ballotData, nvotes int) ([][]int, [][]*types.BigInt) {
+	votes := make([][]int, 0, nvotes)
+	var resultsCh1, resultsCh2 []*types.BigInt
```
Comment (altergui) [inline cmd/end2endtest/helpers.go:454]: it's not immediately clear to me what "choice" ("Ch") means in this var name: are you talking about "Field" or "Value"? (in the terms described in https://developer.vocdoni.io/protocol/ballot)

**Review by altergui — APPROVED** [review]
i made some more comments/suggestions,
but this PR has been in the queue for so long, that i'll prefer to take it as-is and approve it. feel free to merge it without making any changes

File: cmd/end2endtest/helpers.go (hunk 14 above)
Comment (altergui) [reply cmd/end2endtest/helpers.go]: i'd rather meant to move this out of helpers.go , into ballot.go

so keep in helpers.go the code that is used by more than one test (like setupElection), and whatever code is used by just a single test (like ballotVotes) is kept into that specific test (ballot.go)

File: cmd/end2endtest/helpers.go (hunk 15 above)
Comment (mariajdab) [reply cmd/end2endtest/helpers.go:470]: Yes, it's needed, the zero values to convert to the type`[]*types.BigInt` and then sum the valid votes

**Comment by mariajdab** [comment]
Last suggestions were added
